### ISSUE TITLE
Base classes for architecture workhorses

### ISSUE THREAD

**Issue by jbischof** [keras-team/keras-nlp#530]
We essentially have four workhorse classes for each architecture:
1. tokenizer,
2. preprocessor,
3. backbone, and
4. task.

Currently only tokenizers have a their own base class. However, the other three would benefit from this as well because methods like `from_config`, `from_preset`, and eventually `generate` do not seem to be architecture dependent. We might have to slightly change the implementation of some methods (e.g., extracting `backbone_class` for the second and `token_probability_fn` for the third) but our code should become a lot cleaner.

For example, if we look at `BertBackbone.from_preset` ([link](https://github.com/keras-team/keras-nlp/blob/405f1f9737295bda0d79d5426b928ee2623cf3fe/keras_nlp/models/bert/bert_models.py#L263)), we see that there is nothing particular to this architecture in the implementation.

Proposal class hierarchy in BERT example:
```
├── keras.Layer     
|   └ keras_nlp.tokenizers.Tokenizer (already exists)
|       └ keras_nlp.tokenizers.WordPieceTokenizer (already exists, but will own `from_preset`)
|           └ keras_nlp.models.BertTokenizer                 
|   └ keras_nlp.models.Preprocessor
|       └ keras_nlp.models.BertPreprocessor
└── keras.Model                  
    ├── keras_nlp.utils.pipeline_model.PipelineModel
    |    └ Task
    |         ├── keras_nlp.models.BertClassifier   
    |         ├── keras_nlp.models.BertSpanLabeler
    |         └ ...   
    └ keras_nlp.models.Backbone
         └ keras_nlp.models.BertBackbone
```

EDIT (2023-01-11): Removed smurf naming and added tokenizers.

**Comment by mattdangerw** [comment]
Is this the same as #503? Should we close one of these issues to consolidate discussion/

**Comment by jbischof** [comment]
Yes, closed #503, this proposal is more comprehensive and has more detail

**Comment by mattdangerw** [comment]
Thanks!

Will think through this more, but a quick take.

 - I like the idea if we of splitting off common implementation if it make our model code simpler.
 - I don't think we should expose them (do not document or stick them in `__init__.py` files) unless we are expecting users to interact with the symbols directly. This will fit with our other "base classes" in Keras (tokenizer, layer, model, callback, optimizer, etc). The ones we are expose are meant to be sub-classed directly.

So for now, my take would be let's do this if we think it would improve our internal organization of the models directory. And take whatever factorization lightly, be willing to continue to tweak it heavily as we add decode and sequence to sequence models.

**Comment by fchollet** [comment]
What functionality does `BackboneModel` implement compared to `Model`?

> I don't think we should expose them (do not document or stick them in __init__.py files) unless we are expecting users to interact with the symbols directly. This will fit with our other "base classes" in Keras (tokenizer, layer, model, callback, optimizer, etc). The ones we expose are meant to be sub-classed directly.

Yes. We'd only expose classes we expect you users to either instantiate directly, or subclass. In core Keras we never got around to exposing the base preprocessing layer, for instance, since users aren't creating their own in practice.

If we only expect contributors to the repo to subclass these classes, they don't need to be public.

One nitpick: I'd generally favor avoiding Smurf style naming, e.g. naming layers `xxxLayer` (unless the fact that it's a layer needs to be restated). `Preprocessor` seems cleaner, as long as it's not ambiguous.

**Comment by chenmoneygithub** [comment]
We may need to think more about how different models can share the same abstract class and what benefits it brings. For `BackboneModel`, it has `from_presets()` that can be shared across models.  For `TaskModel` I am not sure, since tasks have a high variance, even if we have a base class, I expect child classes to implement many of their own methods, so the abstraction is less useful than `PipelineModel`, which simply provides the preprocessing functionality.

**Comment by jbischof** [comment]
In flight PRs and issues:
#648, #635, #636

**Comment by jbischof** [comment]
This issue is now resolved with #671 and #673!
